### PR TITLE
Tabular: Added persist_models, unpersist_models methods to TabularPredictor

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*
#376 

*Description of changes:*

Added persist_models, unpersist_models methods to TabularPredictor

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-640-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-640/1/index.html

**Comment by szha** [comment]
Job PR-640-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-640/2/index.html

File: autogluon/task/tabular_prediction/predictor.py [hunk 1]
```diff
@@ -622,6 +622,49 @@ def feature_importance(self, dataset=None, model=None, features=None, feature_st
...
+
+        Parameters
+        ----------
+        models : list of str or str, default = None
+            Model names of models to persist.
+            If None then all models are persisted.
+            If 'best' then the model with the highest validation score is persisted.
+            Valid models are listed in this `predictor` by calling `predictor.get_model_names()`.
+        with_ancestors : bool, default = True
+            If True, all ancestor models of the provided models will also be persisted.
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/predictor.py:638]: Clarify what downsides arise if False. Also may want to clarify setting where ancestors are even relevant (stacking).

File: autogluon/task/tabular_prediction/predictor.py [hunk 2]
```diff
@@ -622,6 +622,49 @@ def feature_importance(self, dataset=None, model=None, features=None, feature_st
...
+        """
+        Persist models in memory for reduced inference latency. This is particularly important if the models are being used for online-inference where low latency is critical.
+        If models are not persisted in memory, they are loaded from disk every time they are asked to make predictions.
+
+        Parameters
+        ----------
+        models : list of str or str, default = None
+            Model names of models to persist.
+            If None then all models are persisted.
+            If 'best' then the model with the highest validation score is persisted.
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/predictor.py:635]: Shouldn't the default be 'best'?  I imagine this `persist_models()` is primarily used before inference

File: autogluon/task/tabular_prediction/predictor.py [hunk 3]
```diff
@@ -622,6 +622,49 @@ def feature_importance(self, dataset=None, model=None, features=None, feature_st
...
+            Valid models are listed in this `predictor` by calling `predictor.get_model_names()`.
+        with_ancestors : bool, default = True
+            If True, all ancestor models of the provided models will also be persisted.
+        max_memory : float, default = 0.1
+            Proportion of total available memory to allow for the persisted models to use.
+            If the models' summed memory usage requires a larger proportion of memory than max_memory, they are not persisted. In this case, the output will be an empty list.
+            If None, then models are persisted regardless of estimated memory usage. This can cause out-of-memory errors.
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/predictor.py:642]: Maybe point user to chunked inference trick here as well?

File: autogluon/task/tabular_prediction/predictor.py [hunk 4]
```diff
@@ -622,6 +622,49 @@ def feature_importance(self, dataset=None, model=None, features=None, feature_st
...
+        Returns
+        -------
+        List of persisted model names.
+        """
+        return self._learner.persist_trainer(low_memory=False, models=models, with_ancestors=with_ancestors, max_memory=max_memory)
+
+    def unpersist_models(self, models=None) -> list:
+        """
+        Unpersist models in memory for reduced memory usage.
+        If models are not persisted in memory, they are loaded from disk every time they are asked to make predictions.
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/predictor.py:653]: Should we clarify another way to reset the predictor / remove its models from memory is simply to reload it?

File: autogluon/task/tabular_prediction/predictor.py (hunk 2 above)
Comment (jwmueller) [reply autogluon/task/tabular_prediction/predictor.py:635]: May also clarify: "If 'best' then the model with the highest validation score is persisted (this is the model used for prediction by default)."

**Comment by szha** [comment]
Job PR-640-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-640/3/index.html

File: autogluon/utils/tabular/ml/trainer/abstract_trainer.py [hunk 5]
```diff
@@ -983,27 +984,60 @@ def save_model(self, model, reduce_memory=True):
...
-    def load_models_into_memory(self, model_names=None):
+    def persist_models(self, model_names: list = None, with_ancestors=False, max_memory=None) -> list:
         if model_names is None:
             model_names = self.get_model_names_all()
+        elif model_names is 'best':
+            if self.model_best is not None:
+                model_names = [self.model_best]
+            else:
+                model_names = [self.get_model_best(can_infer=True)]
+        if with_ancestors:
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/trainer/abstract_trainer.py:1002]: Should there be a check that `model_names` is valid argument before this moving on with this line?

File: autogluon/task/tabular_prediction/predictor.py (hunk 1 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/predictor.py:638]: added

File: autogluon/task/tabular_prediction/predictor.py (hunk 2 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/predictor.py:635]: added

File: autogluon/task/tabular_prediction/predictor.py (hunk 3 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/predictor.py:642]: I think this is better saved for a tutorial, they aren't very related.

File: autogluon/task/tabular_prediction/predictor.py (hunk 4 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/predictor.py:653]: added

File: autogluon/utils/tabular/ml/trainer/abstract_trainer.py (hunk 5 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/trainer/abstract_trainer.py:1002]: added

**Comment by Innixma** [comment]
> Looks good to me, just have minor comments about clarity. I also think 'best' should be the default for models in `predictor.persist_models()` since most users probably don't require all other models persisted as well.
> 
> Should also consider if we want to persist at the end of `task.fit()` by default or not. I imagine if the models occupy hardly any of the available memory, then persisting them by default is reasonable.

Regarding persisting by default, this is planned for a future PR to ensure stability, since this is a major change in functionality and we need to be careful about the % of memory to allow by default.

**Comment by szha** [comment]
Job PR-640-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-640/4/index.html

**Pull request merged**